File: gad-rust/src/uuu/parse/amen.rs
//
// amen.rs
//
use std::fs::File;
use std::io::Write;

use crate::uuu::parse::vars::*;

pub fn gen_loop(mut f :&File, gen :&str, ident :i32 ) -> i32 {
  
  wr(f,b"\n"); let ii :i32 = ident - 2; to(f,ii);
  match gen {
  "-rust" | "-go" => { let _ = wr(f, b"};\n"); },
  "-mojo" | "-python" => { let _ = wr(f, b"pass\n"); },
  _ => { },
  }; 
  return ii;
}

pub fn gen_done(mut f :&File, gen :&str, ident :i32 ) -> i32 { return gen_loop( f, gen, ident ); }

pub fn gen_amen(mut f :&File, gen :&str, ident :i32 ) -> i32 {
  
  set_in_proc(false); 
  wr(f,b"\n"); let ii :i32 = ident - 2; to(f,ii);
  match gen {
  "-rust" | "-go" => { let _ = wr(f, b"}\n"); },
  "-mojo" | "-python" => { let _ = wr(f, b"pass\n"); },
  _ => { },
  };
  
  return ii;
}

File: gad-go/amen.go
//
// amen.go
//
package main

func EvClear() {
}

func GenInit() {
  Wr("\n")
  InArray = false;
}


func RustLoop() {
  To(GetIdent()); Wr("}\n");
}

func GoLoop() {
  To(GetIdent()); Wr("}\n");
}

func GenLoop() {
  DbgTrace("GenLoop");
  switch Mode {
  case ASM32: {
    if Nev > 0 { Asm32Loop(); }; 
    return;
  }
  case ASM: { 
    if Nev > 0 { AsmLoop(); }; 
    return; 
  };
  };
  SetIdent(GetIdent() - 2); To(GetIdent());
  switch Mode {
  case RUST: RustLoop();
  case GO:   GoLoop();
  case MOJO: MojoLoop();
  case PYTHON: PyLoop();
  };
}

func GenDone() {
  switch Mode { 
  case ASM32: { Asm32Done(); return; } 
  case ASM: { AsmDone(); return; }
  };
  GenLoop(); 
}

func Asm32Amen() {
  InProc = false;
  Wr("# amen ",CurProc,"\n");
  Wr("  pop %eax\n","  ret\n");
}

func AsmAmen() {
  InProc = false;
  Wr("# amen ",CurProc,"\n");
  Wr("  pop %rax\n","  ret\n");
}

func GenAmen() {
  switch Mode {
  case ASM32: { Asm32Amen(); return; }
  case ASM: { AsmAmen(); return;  }
  };
  InProc = false; 
  To(GetIdent() - 2);
  switch Mode {
  case RUST, GO: { Wr("}\n"); }
  case MOJO,PYTHON: { Wr("pass\n"); }
  default:
  };
}

